File: String/P49GroupAnagrams/GroupAnagrams.py
import unittest
from typing import List
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        map = defaultdict(dict)
        for s in strs:
            temp = s
            ch = sorted(list(s))
            sortedStr = "".join(ch)
            if sortedStr not in map:
                map[sortedStr] = []
            map[sortedStr].append(temp)
        res = []
        for value in map.values():
            res.append(value)
        return res

class UnitTest(unittest.TestCase):
    def test(self):
        strs = ["eat","tea","tan","ate","nat","bat"]
        self.assertEqual([["eat","tea","ate"],["tan","nat"],["bat"]], Solution().groupAnagrams(strs))
if __name__ == "__main__":
    unittest.main()

File: String/P49GroupAnagrams/GroupAnagrams.java
package String.P49GroupAnagrams;

import java.util.*;

/**
 * @CLassName GroupAnagrams
 * @Description
 * Given an array of strings strs, group the anagrams together. You can return the answer in any order.
 *
 * An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.
 *
 * Example 1:
 *
 * Input: strs = ["eat","tea","tan","ate","nat","bat"]
 * Output: [["bat"],["nat","tan"],["ate","eat","tea"]
 *
 * @Author cst
 * @Date 2/3/21  1:49 PM
 */
public class GroupAnagrams {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> res = new ArrayList<>();
        Map<String, List<String>> map = new HashMap<>();
        for (String str : strs) {
            String temp = str;
            char[] ch = str.toCharArray();
            Arrays.sort(ch);
            System.out.println(new String(ch));
            if(!map.containsKey(new String(ch))) {
                map.put(new String(ch), new ArrayList<>());
            }
            map.get(new String(ch)).add(temp);
        }
        res.addAll(map.values());
        return res;
    }
}
